### PR TITLE
Documentation workflow context apply

### PR DESCRIPTION
**Pull request by noniagriconomie**
Hi,

While attending to the #SymfonyConHackday2018

I wanted to test @lyrixx implementation (https://github.com/symfony/symfony/pull/29146) of a request of mine (https://github.com/symfony/symfony/issues/27925) on one of my projects.

I also wanted to add/improve the documentation in the same time.

This PR is WIP as the feature is not merged yet (on its way)

thank you

### CODE REVIEW

File: workflow.rst [hunk 1]
```diff
@@ -8,8 +8,9 @@ best kept away from your models and should be defined in configuration.
 
 A **definition** of a workflow consist of places and actions to get from one
 place to another. The actions are called **transitions**. A workflow does also
-need to know each object's position in the workflow. That **marking store** writes
-to a property of the object to remember the current place.
+need to know each object's position in the workflow. This is the goal of the
+**marking store** that writes to a property of the object the current **place(s)**
```
Comment (lyrixx) [inline workflow.rst]: Technically speaking, the marking store could store the places **anywhere**: an API, a file, a table ...
I think we could replace `to a property` by `to a property or somewhere else`.

File: workflow/state-machines.rst [hunk 2]
```diff
@@ -209,6 +209,33 @@ you can get this state machine by injecting the Workflow registry service::
...
         {
             $stateMachine = $this->workflows->get($subject, 'pull_request');
+            $stateMachine->apply($subject, 'wait_for_review');
+            // ...
+        }
+
+        // ...
+    }
+
+You can also directly inject the proper :class:`Symfony\\Component\\Workflow\\StateMachine`
```
Comment (lyrixx) [inline workflow/state-machines.rst]: Do it works out of the box ?
I will work only if you have one and only one StateMachine?
In my project, I had to do that to make it work:

```yaml
services:
    # We have only one state machine, so we can be lazy
    Symfony\Component\Workflow\StateMachine:
        alias: state_machine.order
```

**Review by lyrixx — COMMENTED** [review]
Thanks a lot for working on this feature :+1:

File: workflow.rst (hunk 1 above)
Comment (noniagriconomie) [reply workflow.rst]: understood

File: workflow/state-machines.rst (hunk 2 above)
Comment (noniagriconomie) [reply workflow/state-machines.rst]: Are you speaking of autowiring?
Actually in many of my projects i am still on 3.4 lts without using autowire, so when i wrote the StateMachine i also wrote the service id

but yes in case of autowiring actuivated, with ≥ 2 StateMachine, we need to create aliases

I do not know how to properly document such different use case depending on the code

**Comment by noniagriconomie** [comment]
@lyrixx review addressed

File: workflow/usage.rst [hunk 3]
```diff
@@ -43,7 +43,7 @@ like this:
                     audit_trail:
                         enabled: true
                     marking_store:
-                        type: 'multiple_state' # or 'single_state'
+                        type: 'multiple_state' # or 'single_state', 'method'
```
Comment (lyrixx) [inline workflow/usage.rst]: method was added in 4.3. we should be clear about that.

**Review by lyrixx — COMMENTED** [review]
:+1:

File: workflow/usage.rst (hunk 3 above)
Comment (noniagriconomie) [reply workflow/usage.rst]: What is the best way to write this?
In a comment next to this value ?
Above all this code ?

**Comment by noniagriconomie** [comment]
hi @lyrixx 

I do not know if the comment like that is ok (it is a bit weird), wdyt? 

Also, do you have a date for the merge of the related feature ? :)

Thank you!

**Comment by javiereguiluz** [comment]
@noniagriconomie I'd like to merge your pull request, but I'm seeing lots of conflicts when doing that. Could you please rebase this to the latest changes of the master branch? Thanks!

**Comment by noniagriconomie** [comment]
@javiereguiluz of course i will do this asap

**Comment by noniagriconomie** [comment]
@javiereguiluz Done

Just i am not sure about the "versionadded:: 4.1" part when i rebased

thank you

**Review by HeahDude — APPROVED** [review]
Thanks!

File: workflow/state-machines.rst [hunk 4]
```diff
@@ -209,6 +209,35 @@ you can get this state machine by injecting the Workflow registry service::
...
             $stateMachine = $this->workflows->get($subject, 'pull_request');
+            $stateMachine->apply($subject, 'wait_for_review');
+            // ...
+        }
+
+        // ...
+    }
+
+Symfony also creates automatically for you a service for each workflow (:class:`Symfony\\Component\\Workflow\\Workflow`) or state machine (:class:`Symfony\\Component\\Workflow\\StateMachine`) you have defined in your configuration.
+This means that you can use respectively ``workflow.pull_request`` or ``state_machine.pull_request`` in your service definition to have directly the proper service.
```
Comment (OskarStark) [inline workflow/state-machines.rst]:
```suggestion
This means that you can use respectively ``workflow.pull_request`` or
``state_machine.pull_request`` in your service definition to have directly
the proper service::
```

File: workflow/state-machines.rst [hunk 5]
```diff
@@ -209,6 +209,35 @@ you can get this state machine by injecting the Workflow registry service::
...
+
```
Comment (OskarStark) [inline workflow/state-machines.rst]:
```suggestion

```

File: workflow/state-machines.rst [hunk 6]
```diff
@@ -209,6 +209,35 @@ you can get this state machine by injecting the Workflow registry service::
...
+::
```
Comment (OskarStark) [inline workflow/state-machines.rst]:
```suggestion

```

File: workflow/state-machines.rst [hunk 7]
```diff
@@ -209,6 +209,35 @@ you can get this state machine by injecting the Workflow registry service::
...
+
+        public function __construct(StateMachine $stateMachine)
+        {
+            $this->stateMachine = $stateMachine;
+        }
+
+        public function someMethod($subject)
+        {
+            $this->stateMachine->apply($subject, 'wait_for_review', [
+                'log_comment' => 'My logging comment for the wait for review transition.'
```
Comment (OskarStark) [inline workflow/state-machines.rst]:
```suggestion
                'log_comment' => 'My logging comment for the wait for review transition.',
```

File: workflow/usage.rst [hunk 8]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
                 // ... if the transition is not allowed
             }
 
+            // Update the currentState on the post passing some contextual data
+            // to the whole workflow process
+            try {
+                $workflow->apply($post, 'publish', [
+                    'log_comment' => 'My logging comment for the publish transition.'
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion
                    'log_comment' => 'My logging comment for the publish transition.',
```

File: workflow/usage.rst [hunk 9]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
...
+            } catch (TransitionException $exception) {
+                // ... if the transition is not allowed
+            }
+
             // See all the available transitions for the post in the current state
             $transitions = $workflow->getEnabledTransitions($post);
         }
     }
 
+.. versionadded:: 4.1
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion
.. versionadded:: 4.1

```

File: workflow/usage.rst [hunk 10]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
...
+    The :class:`Symfony\\Component\\Workflow\\Exception\\TransitionException`
+    class was introduced in Symfony 4.1.
+
+.. versionadded:: 4.1
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion
.. versionadded:: 4.1

```

File: workflow/usage.rst [hunk 11]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
...
+    The :method:`Symfony\\Component\\Workflow\\Registry::all` method was
+    introduced in Symfony 4.1.
+
+.. versionadded:: 4.3
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion
.. versionadded:: 4.3

```

File: workflow/usage.rst [hunk 12]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
...
+    The :method:`Symfony\\Component\\Workflow\\Workflow::apply` has now a new parameter ``$context``
+    that is passed to the :class:`Symfony\\Component\\Workflow\\MarkingStore\\MarkingStoreInterface`
+    :method:`Symfony\\Component\\Workflow\\MarkingStore\\MarkingStoreInterface::setMarking` method.
+
+An example of usage with the ``$context`` parameter can be when you need, in addition of marking your object
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion
An example of usage with the ``$context`` parameter can be when you need,
in addition of marking your object in its new place, to contextualize this change.
```

File: workflow/usage.rst [hunk 13]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
...
+in its new place, to contextualize this change.
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion

```

File: workflow/usage.rst [hunk 14]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
...
+
+.. tip::
+
+    Configure the ``type`` as ``method`` of the ``marking_store`` option to use this feature
+    without implementing your own marking store (``method`` was added in 4.3).
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion
    without implementing your own marking store.
```

File: workflow/usage.rst [hunk 15]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
...
+
+You can also use this ``$context`` in your own marking store implementation.
+A simple implementation example is when you want to store the place as integer instead of string in your object.
+
+Lets say your object has a status property, stored as an integer in your storage, and you want to log an optional
+comment any time the status changes.
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion
comment any time the status changes::
```

File: workflow/usage.rst [hunk 16]
```diff
@@ -221,11 +221,86 @@ you can get the workflow by injecting the Workflow registry service::
...
+
+.. code-block:: php
```
Comment (OskarStark) [inline workflow/usage.rst]:
```suggestion

```

File: workflow/usage.rst [hunk 17]
```diff
@@ -234,6 +244,63 @@ you can get the workflow by injecting the Workflow registry service::
...
+    public function setStatus(int $status, ?string $comment = null)
+    {
+        $this->status = $status;
+        $this->addStatusLogRecord(new StatusLog($this, $comment));
+
+        return $this;
+    }
+
+    // the StatusLog class can have a createdAt, a username,
+    // the new status, and finally your optional comment retrived from the workflow context.
```
[the remark this answers is not in the document]
Comment (OskarStark) [reply workflow/usage.rst]: You mean `retrieved` @HeahDude ? 🤣

File: workflow.rst [hunk 18]
```diff
@@ -8,8 +8,9 @@ best kept away from your models and should be defined in configuration.
 
 A **definition** of a workflow consist of places and actions to get from one
 place to another. The actions are called **transitions**. A workflow does also
-need to know each object's position in the workflow. That **marking store** writes
-to a property of the object to remember the current place.
+need to know each object's position in the workflow. This is the goal of the
```
Comment (OskarStark) [inline workflow.rst]:
```suggestion
need to know each position an object can be in the workflow. This is the goal of the
```

What do you think about my rewording?

File: contributing/documentation/overview.rst [hunk 19]
```diff
@@ -270,6 +270,8 @@ purposes following these steps:
 
 The generated documentation is available in the ``_build/html`` directory.
 
+You can also use Docker that wraps all this for you, please visit https://github.com/symfony/symfony-docs#docker.
```
Comment (OskarStark) [inline contributing/documentation/overview.rst]:
```suggestion
.. tip::

    You can also use `Docker`_ that wraps all this for you!
```

And please add this to the end of the file:
```rst
.. _`Docker`: https://github.com/symfony/symfony-docs#docker
```

**Review by OskarStark — COMMENTED** [review]
I left some minor comments!

Great contribution so far 💪 thank you!

**Comment by noniagriconomie** [comment]
@OskarStark (@HeahDude) reviews addressed, thank you!

**Review by lyrixx — APPROVED** [review]
:+1: Very cool. Thanks for your work.

**Comment by Simperfit** [comment]
@xabbuh can we merge this one ?

File: workflow/state-machines.rst [hunk 20]
```diff
@@ -209,6 +209,33 @@ you can get this state machine by injecting the Workflow registry service::
...
         {
             $stateMachine = $this->workflows->get($subject, 'pull_request');
+            $stateMachine->apply($subject, 'wait_for_review');
+            // ...
+        }
+
+        // ...
+    }
+
+Symfony also creates automatically for you a service for each workflow (:class:`Symfony\\Component\\Workflow\\Workflow`) or state machine (:class:`Symfony\\Component\\Workflow\\StateMachine`) you have defined in your configuration.
```
Comment (OskarStark) [inline workflow/state-machines.rst]: I would reword it:

Symfony creates a service for you automatically....

File: workflow/usage.rst [hunk 21]
```diff
@@ -221,11 +221,87 @@ you can get the workflow by injecting the Workflow registry service::
...
+            }
+
             // See all the available transitions for the post in the current state
             $transitions = $workflow->getEnabledTransitions($post);
         }
     }
 
+.. versionadded:: 4.1
+
+    The :class:`Symfony\\Component\\Workflow\\Exception\\TransitionException`
```
Comment (OskarStark) [inline workflow/usage.rst]: Do we Need this versionadded ?

@symfony/team-symfony-docs

File: workflow/usage.rst [hunk 22]
```diff
@@ -221,11 +221,87 @@ you can get the workflow by injecting the Workflow registry service::
...
+    class was introduced in Symfony 4.1.
+
+.. versionadded:: 4.1
+
+    The :method:`Symfony\\Component\\Workflow\\Registry::all` method was
```
Comment (OskarStark) [inline workflow/usage.rst]: Do we Need this versionadded ?

@symfony/team-symfony-docs

**Comment by xabbuh** [comment]
Thank you @noniagriconomie and congratulations on your first contribution to the Symfony documentation. I have merged your PR (I have squased the commits while merging that's why the PR is shown as closed instead of merged) and made some small tweaks in e1e4efcd707cfa7489d6df4ba47a6d1be4a308ea.

File: workflow/state-machines.rst (hunk 20 above)
Comment (xabbuh) [reply workflow/state-machines.rst]: see e1e4efcd707cfa7489d6df4ba47a6d1be4a308ea :)

Comment (OskarStark) [reply workflow/state-machines.rst]: Thanks
